rpcs3/rpcs3qt/main_window.cpp

```diff
@@ -850,29 +874,37 @@ void main_window::HandlePupInstallation(QString file_path)
 
 	const std::string cur_version = "4.87";
 
-	if (version_string < cur_version &&
-		QMessageBox::question(this, tr("RPCS3 Firmware Installer"), tr("Old firmware detected.\nThe newest firmware version is %1 and you are trying to install version %2\nContinue installation?").arg(qstr(cur_version), qstr(version_string)),
+	if (mw && version_string < cur_version &&
+		QMessageBox::question(mw, tr("RPCS3 Firmware Installer"), tr("Trying to install old firmware.\nThe newest firmware version is %1 and you are trying to install version %2\nContinue installation?").arg(qstr(cur_version), qstr(version_string)),
 			QMessageBox::Yes | QMessageBox::No, QMessageBox::Yes) == QMessageBox::No)
 	{
-		return;
+		return false;
 	}
 
+	if (std::string installed = utils::get_firmware_version(); !installed.empty())
+	{
+		if (mw && QMessageBox::question(mw, tr("RPCS3 Firmware Installer"), tr("Firmware of version %1 has already been installed.\nOverwrite current installation with verion %2?").arg(qstr(installed), qstr(version_string)),
+			QMessageBox::Yes | QMessageBox::No, QMessageBox::Yes) == QMessageBox::No)
+		{
+			return false;
+		}
+	}
+
+	if (mw) mw->m_gui_settings->SetValue(gui::fd_install_pup, QFileInfo(file_path).path());
+
 	// Remove possibly PS3 fonts from database
 	QFontDatabase::removeAllApplicationFonts();
 
-	progress_dialog pdlg(tr("RPCS3 Firmware Installer"), tr("Installing firmware version %1\nPlease wait...").arg(qstr(version_string)), tr("Cancel"), 0, static_cast<int>(update_filenames.size()), false, this);
-	pdlg.show();
-
 	// Synchronization variable
-	atomic_t<int> progress(0);
+	atomic_t<uint> progress(0);
+
+	do
```

Comment:
what's the issue with the existing code ?

Reply:
Removed.